numpy/lib/format.py

```diff
@@ -1011,3 +1011,27 @@ def isfileobj(f):
         return True
     except OSError:
         return False
+
+
+class NumpyUnpickler(pickle.Unpickler):
+    """
+    A thin wrapper for :py:class:`pickle.Unpickler` that
+    allows to load 1.x array pickles with numpy 2.0 and
+    2.0 array pickles with numpy 1.26.
+    """
+
+    def find_class(self, module: str, name: str) -> object:
+        import numpy as np
+        numpy_major_version: int = np.lib.NumpyVersion(np.__version__).major
+        if (
+            module.startswith("numpy.core") and
+            numpy_major_version >= 2 and
+            hasattr(np, "_core")
+        ):
+            module = module.replace("core", "_core", 1)
+        elif (
+            module.startswith("numpy._core") and
+            numpy_major_version <= 1
+        ):
+            module = module.replace("_core", "core", 1)
```

Comment:
This is a little hard to read right now and is also doing unnecessary work. You only need to see if the path starts with `np.core` once. Also it's a little weird to have numpy version checks inside numpy. @charris do you think @mtsokol should do a separate PR manually backporting this with a slightly different implementation?

Reply:
I understand it this way: We would like a `NumpyUnpickler` that is able to: 
1. Read 1.x pickles when numpy 2.0 is present
2. Read 2.0 pickles when numpy 1.x is present

In this implementation, if the pickle contains `numpy.core` and numpy 2.0 is installed, we need to replace `core` with `_core` (thus we need to check numpy version). 
If numpy 1.x is present then the path should be left untouched.
If pickle contains `numpy._core` and numpy 1.x is present then we need to change `_core` to `core` so that pickle can be loaded.

I think importing numpy is important here to be aware of the installed version and make path updates based on that.

Reply:
We're inside the numpy source code so we know what version of numpy we're writing code for *a priori* - no need to do the check at runtime. Of course when you backport this code the check will be different, that's why I suggested manually backporting it. Does that make sense?

Reply:
Ah, now I get it. Then I can make separate implementation for 1.x (This implementation is version agnostic - I think two simpler implementations for 1.x and 2.0 are fine for me).